### ISSUE TITLE
Refactor low-level Source operations for better reuse

### ISSUE THREAD

**Issue by sdboyer** [sdboyer/gps#112]
There's a lot of inter-calls on `baseVCSSource`, and its derived types, as they exist today. These types have seen...uh, let's call it "organic growth," and are turning into spaghetti. The main problems are:
- Pretty much all of the methods take out locks - both the "porcelain" and the "plumbing." As a result, porcelain often ends up reimplementing plumbing, as calling the plumbing would deadlock.
- There's a proliferation of individual bits, flags, bools, etc., which represent separate but interrelated states. These need to be consolidated and managed better.

Basically, I got lazy and started focusing on actions, rather than retaining the overall-consistent-state approach, and let things fall into disrepair.

**Comment by sdboyer** [comment]
This is fixed by #196
